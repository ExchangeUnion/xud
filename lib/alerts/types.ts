import { AlertType, ChannelSide } from '../constants/enums';

type BaseAlert = {
  type: AlertType;
  message: string;
  date: number;
};

export type BalanceAlertEvent = {
  /** The total balance of the channel when the alert is triggered. */
  totalBalance: number;
  /** The side of the balance either local or remote. */
  side: ChannelSide;
  /** The balance that triggered the alert. */
  sideBalance: number;
  /** The alert threshold in percentage, e.g. 10 means %10. */
  bound: number;
  /** The currency of the channel. */
  currency: string;
};
export type BalanceAlert = BaseAlert & BalanceAlertEvent;

export type Alert = BalanceAlert;
