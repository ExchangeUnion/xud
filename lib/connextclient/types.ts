/**
 * The configurable options for the connext client.
 */
export type ConnextClientConfig = {
  disable: boolean;
  host: string;
  port: number;
  webhookhost: string;
  webhookport: number;
};

/** General information about the state of this connext client. */
export type ConnextInfo = {
  status: string;
  error?: string;
  address?: string;
  chain?: string;
  version?: string;
};

/** Response for eth_blockNumber call  */
export type ConnextBlockNumberResponse = {
  result: string;
};

/** Response for eth_gasPrice call  */
export type EthproviderGasPriceResponse = {
  result: string;
};

type Balance = {
  amount: string[];
  to: string[];
};

/** Connext channel details */
export type ConnextChannelDetails = {
  assetIds: string[];
  balances: Balance[];
  channelAddress: string;
  merkleRoot: string;
  processedDepositsA: string[];
  processedDepositsB: string[];
  networkContext: {
    chainId: number;
    channelFactoryAddress: string;
    transferRegistryAddress: string;
    providerUrl: string;
  };
  nonce: number;
  alice: string;
  aliceIdentifier: string;
  bob: string;
  bobIdentifier: string;
  timeout: string;
  latestUpdate: {
    assetId: string;
    balance: Balance;
    channelAddress: string;
    details: {
      totalDepositsAlice: string;
      totalDepositsBob: string;
    };
    fromIdentifier: string;
    nonce: number;
    aliceSignature: string;
    bobSignature: string;
    toIdentifier: string;
    type: string;
  };
  defundNonce: string;
  inDispute: boolean;
};

/**
 * The payload for tokenPayment call.
 */
export type ConnextTransferRequest = {
  type: 'HashlockTransfer';
  channelAddress: string;
  amount: string;
  assetId: string;
  details: {
    lockHash: string;
    expiry: string;
  };
  recipient: string;
  meta: {
    routingId: string;
  };
  publicIdentifier: string;
};

/**
 * The response for errors.
 */
export type ConnextErrorResponse = { message: string };

export type ConnextConfig = {
  publicIdentifier: string;
  signerAddress: string;
  index: number;
};

/**
 * The response for /config call.
 */
export type ConnextConfigResponse = ConnextConfig[];

/**
 * The response for /channel call.
 */
export type ConnextChannelResponse = string[];

/**
 * The response for balance call.
 */
export type ConnextChannelBalanceResponse = {
  freeBalanceOffChain: string;
  nodeFreeBalanceOffChain: string;
  freeBalanceOnChain: string;
};

export type GetBlockByNumberResponse = {
  result: {
    difficulty: string;
    extraData: string;
    gasLimit: string;
    gasUsed: string;
    hash: string;
    logsBloom: string;
    miner: string;
    mixHash: string;
    nonce: string;
    number: string;
    parentHash: string;
    receiptsRoot: string;
    sha3Uncles: string;
    size: string;
    stateRoot: string;
    timestamp: string;
    totalDifficulty: string;
    transactions: string[];
    transactionsRoot: string;
    uncles: string[];
  };
};

/**
 * The response for hashLockTransfer call.
 */
export type ConnextTransferResponse = {
  channelAddress: string;
  transferId: string;
  routingId: string;
};

/**
 * The response for withdraw call.
 */
export type ConnextWithdrawResponse = {
  channelAddress: '0xa929dB0530daB525596f5d48Fb5C322fDa8A337B';
  transferId: '0xc2e4592d3fb6c02ee1d3b07bed83b5914f8ca084b0f91d6b80bf8107e58c9c38';
};

type ConnextRoutingPath = {
  recipient: string;
  recipientChainId: number;
  recipientAssetId: string;
};

export type ConnextTransfer = {
  channelFactoryAddress: string;
  assetId: string;
  chainId: number;
  channelAddress: string;
  balance: {
    amount: string[];
    to: string[];
  };
  initiator: string;
  responder: string;
  initialStateHash: string;
  transferDefinition: string;
  transferEncodings: string[];
  transferId: string;
  transferState: {
    lockHash: string;
    expiry: string;
  };
  transferTimeout: string;
  meta: {
    requireOnline: boolean;
    routingId: string;
    path: ConnextRoutingPath[];
    senderIdentifier?: string;
  };
  transferResolver?: {
    preImage?: string;
  };
};

export type TransfersByRoutingIdResponse = ConnextTransfer[];

export type ExpectedIncomingTransfer = {
  rHash: string;
  units: bigint;
  expiry: number;
  tokenAddress: string;
  routingId: string;
};

export type ConnextPreimageRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: string;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      requireOnline: boolean;
      routingId: string;
      path: [
        {
          recipient: string;
          recipientChainId: number;
          recipientAssetId: string;
        },
      ];
    };
    transferResolver: {
      preImage: string;
    };
  };
  conditionType: string;
};

export type ConnextIncomingTransferRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: number;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      routingId: string;
    };
  };
  conditionType: string;
};

export type ConnextDepositConfirmedRequest = {
  id: string;
  data?: {
    amount: {
      _hex: string;
    };
    assetId: string;
    hash: string;
  };
};

export type ProvidePreimageEvent = {
  rHash: string;
  preimage: string;
};

export type TransferReceivedEvent = {
  tokenAddress: string;
  rHash: string;
  expiry: number;
  units: bigint;
  routingId: string;
};
