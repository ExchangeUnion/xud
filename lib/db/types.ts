import { BelongsToGetAssociationMixin, Model } from 'sequelize';
import { ReputationEvent, SwapClientType } from '../constants/enums';
import { Currency, Order, Pair } from '../orderbook/types';
import { Address, NodeConnectionInfo } from '../p2p/types';
import { SwapDeal } from '../swaps/types';

/*
 * The following definitions are in sets of triplets, one for each Model (which represents a table in the database).
 *
 * "xCreationAttributes" is the type definition for the object which is required when a new record is to be created.
 *
 * "xAttributes" is the type definition of the record.
 *
 * "xInstance" is the type definition of a fetched record as a Sequelize row instance, which contains some util properties.
 */

/* Currency */
export type CurrencyCreationAttributes = Currency;

export type CurrencyAttributes = CurrencyCreationAttributes;

export interface CurrencyInstance extends Model<CurrencyAttributes, CurrencyCreationAttributes>, CurrencyAttributes {}

/* SwapDeal */
export type SwapDealCreationAttributes = Pick<
  SwapDeal,
  Exclude<keyof SwapDeal, 'takerMaxTimeLock' | 'price' | 'pairId' | 'isBuy' | 'takerUnits' | 'makerUnits'>
>;

export type SwapDealAttributes = SwapDealCreationAttributes & {
  /** The internal db node id of the counterparty peer for this swap deal. */
  nodeId: number;
};

export interface SwapDealInstance extends Model<SwapDealAttributes, SwapDealCreationAttributes>, SwapDealAttributes {
  getNode: BelongsToGetAssociationMixin<NodeInstance>;
  getOrder: BelongsToGetAssociationMixin<OrderInstance>;
  Node?: NodeInstance;
  Order?: OrderInstance;
}

export type OrderCreationAttributes = Pick<Order, Exclude<keyof Order, 'quantity' | 'hold' | 'price'>> & {
  /** The internal db node id of the peer that created this order. */
  nodeId?: number;
  localId?: string;
  /** The price for the order expressed in units of the quote currency. */
  price?: number;
};

export type OrderAttributes = OrderCreationAttributes & {};

export interface OrderInstance extends Model<OrderAttributes, OrderCreationAttributes>, OrderAttributes {}

export type TradeCreationAttributes = {
  /** The order id of the maker order involved in this trade. */
  makerOrderId: string;
  /** The order id of the taker order involved in this trade, if applicable. */
  takerOrderId?: string;
  /** The rHash of the swap that filled this trade, if applicable. */
  rHash?: string;
  /** The quantity transacted in this trade. */
  quantity: number;
};

export type TradeAttributes = TradeCreationAttributes & {};

export interface TradeInstance extends Model<TradeAttributes, TradeCreationAttributes>, TradeAttributes {
  getMakerOrder: BelongsToGetAssociationMixin<OrderInstance>;
  getTakerOrder: BelongsToGetAssociationMixin<OrderInstance>;
  getSwapDeal: BelongsToGetAssociationMixin<SwapDealInstance>;
  SwapDeal?: SwapDealInstance;
  makerOrder?: OrderInstance;
  takerOrder?: OrderInstance;
  createdAt: Date;
}

/* Node */
export type NodeCreationAttributes = NodeConnectionInfo;

export type NodeAttributes = NodeCreationAttributes & {
  id: number;
  banned: boolean;
  addressesText: string;
  lastAddressText: string;
  lastAddress: Address;
  consecutiveConnFailures: number;
};

export interface NodeInstance extends Model<NodeAttributes, NodeCreationAttributes>, NodeAttributes {
  reputationScore: number;
}

/* Pairs */
export type PairCreationAttributes = Pair;

export type PairAttributes = PairCreationAttributes & {
  id: string;
};

export interface PairInstance extends Model<PairAttributes, PairCreationAttributes>, PairAttributes {}

/* Reputation events */
export type ReputationEventCreationAttributes = {
  event: ReputationEvent;
  nodeId: number;
};

export type ReputationEventAttributes = ReputationEventCreationAttributes & {
  createdAt: number;
  id: number;
};

export interface ReputationEventInstance
  extends Model<ReputationEventAttributes, ReputationEventCreationAttributes>,
    ReputationEventAttributes {}

/* Passwords */
export type PasswordCreationAttributes = {
  encryptedPassword: string;
  currency?: string;
  swapClient: SwapClientType;
};

export type PasswordAttributes = PasswordCreationAttributes & {
  createdAt: number;
  id: number;
};

export interface PasswordInstance extends Model<PasswordAttributes, PasswordCreationAttributes>, PasswordAttributes {}
