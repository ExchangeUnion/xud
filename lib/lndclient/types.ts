/** The configurable options for the lnd client. */

export type LndClientConfig = {
  disable: boolean;
  certpath: string;
  macaroonpath: string;
  host: string;
  port: number;
  nomacaroons: boolean;
  cltvdelta: number;
};

/** General information about the state of this lnd client. */
export type LndInfo = {
  status: string;
  error?: string;
  channels?: ChannelCount;
  chains?: Chain[];
  blockheight?: number;
  uris?: string[];
  version?: string;
  alias?: string;
};

export type ChannelCount = {
  active: number;
  inactive?: number;
  pending: number;
  closed: number;
};

export type Chain = {
  network: string;
  chain: string;
};

export type ClientMethods =
  | 'close'
  | 'waitForReady'
  | 'makeClientStreamRequest'
  | 'makeBidiStreamRequest'
  | 'makeServerStreamRequest'
  | 'makeUnaryRequest';
