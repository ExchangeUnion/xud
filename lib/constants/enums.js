const enums = exports;

enums.swapProtocols = {
  LND: 'LND',
  RAIDEN: 'RAIDEN',
};

enums.orderingDirections = {
  DESC: 'DESC',
  ASC: 'ASC',
};
