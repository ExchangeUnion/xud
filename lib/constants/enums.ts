/** An enumeration of payment channel network clients that support token swaps. */
export enum SwapClients {
  Lnd,
  Raiden,
}

export enum OrderingDirection {
  Desc = 'DESC',
  Asc = 'ASC',
}

export enum OrderSide {
  Buy,
  Sell,
}

export enum Network {
  MainNet = 'mainnet',
  TestNet = 'testnet',
  SimNet = 'simnet',
  RegTest = 'regtest',
}

/**
 * Magic value indicating wire msg origin network, and used to seek to next msg when stream state is unknown
 */
export enum NetworkMagic {
  MainNet = 0xd9b4bef9,
  TestNet = 0x0709110b,
  SimNet = 0x12141c16,
  RegTest = 0xdab5bffa,
}

export enum SwapRole {
  Taker = 0,
  Maker = 1,
}

export enum SwapPhase {
  /** The swap has been created locally. */
  SwapCreated = 0,
  /** The swap has been sent to a peer to request approval. */
  SwapRequested = 1,
  /** The terms of the swap have been agreed to, and we will attempt to execute it. */
  SwapAgreed = 2,
  /**
   * We have commanded lnd to send payment according to the agreed terms. The payment (and swap)
   * could still fail due to no route with sufficient capacity, lack of cooperation from the
   * receiver or any intermediary node along the route, or an unexpected error from lnd.
   */
  SendingAmount = 3,
  /** We have received the agreed amount of the swap, and the preimage is now known to both sides. */
  AmountReceived = 4,
  /** The swap has been formally completed and both sides have confirmed they've received payment. */
  SwapCompleted = 5,
}

export enum SwapState {
  Active = 0,
  Error = 1,
  Completed = 2,
}

export enum ReputationEvent {
  ManualBan = 0,
  ManualUnban = 1,
  PacketTimeout = 2,
  WireProtocolErr = 3,
  InvalidAuth = 4,
  SwapSuccess = 5,
  /** When a swap is accepted and is attempted to be executed but fails. */
  SwapFailure = 6,
  /** When a swap is accepted and then fails due to exceeding time limits. */
  SwapTimeout = 7,
  /** When a swap fails due to unexpected or possibly malicious behavior. */
  SwapMisbehavior = 8,
}

export enum SwapFailureReason {
  /** Could not find the order specified by a swap request. */
  OrderNotFound = 0,
  /** The order specified by a swap request is on hold for a different ongoing swap. */
  OrderOnHold = 1,
  /** The swap request contained invalid data. */
  InvalidSwapRequest = 2,
  /** We are not connected to both swap clients, or we are missing pub key identifiers for the peer's nodes. */
  SwapClientNotSetup = 3,
  /** Could not find a route to complete the swap. */
  NoRouteFound = 4,
  /** A call to lnd failed for an unexpected reason. */
  UnexpectedLndError = 5,
  /** Received a swap packet from the peer with invalid data. */
  InvalidSwapPacketReceived = 6,
  /** The call to lnd to send payment failed. */
  SendPaymentFailure = 7,
  /** The swap resolver request from lnd was invalid. */
  InvalidResolveRequest = 8,
  /** The swap request attempts to reuse a payment hash. */
  PaymentHashReuse = 9,
  /** The swap timed out while we were waiting for it to complete execution. */
  SwapTimedOut = 10,
  /** The deal timed out while we were waiting for the peer to respond to our swap request. */
  DealTimedOut = 11,
  /** The swap failed due to an unrecognized error. */
  UnknownError = 12,
}

export enum DisconnectionReason {
  ResponseStalling = 1,
  IncompatibleProtocolVersion = 2,
  UnexpectedIdentity = 3,
  ForbiddenIdentityUpdate = 4,
  ConnectedToSelf = 5,
  NotAcceptingConnections = 6,
  Banned = 7,
  AlreadyConnected = 8,
  Shutdown = 9,
  MalformedVersion = 10,
  AuthFailureInvalidTarget = 11,
  AuthFailureInvalidSignature = 12,
  WireProtocolErr = 13,
}
