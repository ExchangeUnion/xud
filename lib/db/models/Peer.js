module.exports = (sequelize, DataTypes) => {
  const Peer = sequelize.define('Peer', {
    id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
    address: { type: DataTypes.STRING(39), allowNull: false },
    port: { type: DataTypes.INTEGER, allowNull: false },
    pubKey: { type: DataTypes.STRING, allowNull: true, unique: true },
  }, {
    tableName: 'peers',
    indexes: [{
      unique: true,
      fields: ['address', 'port'],
    }],
  });

  return Peer;
};
