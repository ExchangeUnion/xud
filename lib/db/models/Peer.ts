import Sequelize from 'sequelize';
import { db } from '../../types';

export default (sequelize: Sequelize.Sequelize, DataTypes: Sequelize.DataTypes) => {
  const attributes: db.SequelizeAttributes<db.PeerAttributes> = {
    id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
    address: { type: DataTypes.STRING(39), allowNull: false },
    port: { type: DataTypes.INTEGER, allowNull: false },
    pubKey: { type: DataTypes.STRING, allowNull: true, unique: true },
  };

  const indexes: Sequelize.DefineIndexesOptions[] = [{
    unique: true,
    fields: ['address', 'port'],
  }];

  const options: Sequelize.DefineOptions<db.PeerInstance> = {
    indexes,
    tableName: 'hosts',
  };

  const Peer = sequelize.define<db.PeerInstance, db.PeerAttributes>('Peer', attributes, options);

  return Peer;
};
