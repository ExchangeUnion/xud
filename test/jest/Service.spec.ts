import { Owner, SwapClientType } from '../../lib/constants/enums';
import Logger from '../../lib/Logger';
import Orderbook from '../../lib/orderbook/OrderBook';
import Peer from '../../lib/p2p/Peer';
import Pool from '../../lib/p2p/Pool';
import Service from '../../lib/service/Service';
import { ServiceComponents } from '../../lib/service/types';
import SwapClient from '../../lib/swaps/SwapClient';
import SwapClientManager from '../../lib/swaps/SwapClientManager';
import Swaps from '../../lib/swaps/Swaps';
import NodeKey from '../../lib/nodekey/NodeKey';

jest.mock('../../lib/nodekey/NodeKey');
const mockedNodeKey = <jest.Mock<NodeKey>>(<any>NodeKey);
jest.mock('../../lib/orderbook/OrderBook');
const mockedOrderbook = <jest.Mock<Orderbook>>(<any>Orderbook);
jest.mock('../../lib/swaps/Swaps');
const mockedSwaps = <jest.Mock<Swaps>>(<any>Swaps);
jest.mock('../../lib/swaps/SwapClientManager', () => {
  return jest.fn().mockImplementation(() => {
    return {
      getType: () => SwapClientType.Lnd,
      getOutboundReservedAmount: () => 0,
    };
  });
});
const mockedSwapClientManager = <jest.Mock<SwapClientManager>>(<any>SwapClientManager);
jest.mock('../../lib/swaps/SwapClient');
const mockedSwapClient = <jest.Mock<SwapClient>>(<any>SwapClient);
jest.mock('../../lib/p2p/Pool');
const mockedPool = <jest.Mock<Pool>>(<any>Pool);
jest.mock('../../lib/p2p/Peer', () => {
  return jest.fn().mockImplementation(() => {
    return {
      getLndPubKey: () => lndPubKey,
      getIdentifier: () => lndPubKey,
      getLndUris: jest.fn(),
    };
  });
});
const mockedPeer = <jest.Mock<Peer>>(<any>Peer);
jest.mock('../../lib/Logger');
const mockedLogger = <jest.Mock<Logger>>(<any>Logger);

const lndPubKey = 'lndpubkey';

const getArgs = () => {
  return {
    nodeIdentifier: '02f8895eb03c37b2665415be4d83b20228acc0abc55ebf6728565141c66cfc164a',
    amount: 16000000,
    currency: 'BTC',
  };
};
describe('Service', () => {
  let service: Service;
  let components: ServiceComponents;
  let peer: Peer;

  beforeEach(() => {
    components = {
      orderBook: new mockedOrderbook(),
      swapClientManager: new mockedSwapClientManager(),
      pool: new mockedPool(),
      swaps: new mockedSwaps(),
      version: '1.0.0',
      shutdown: jest.fn(),
      nodeKey: new mockedNodeKey(),
      logger: new mockedLogger(),
    };
    peer = new mockedPeer();
    components.pool.getPeer = jest.fn().mockReturnValue(peer);
    components.swapClientManager.openChannel = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('openChannel', () => {
    test('gets peer identifier from pool for swapClientManager', async () => {
      service = new Service(components);
      const args = getArgs();
      await service.openChannel(args);
      expect(components.pool.getPeer).toHaveBeenCalledWith(args.nodeIdentifier);
      expect(components.swapClientManager.openChannel).toHaveBeenCalledWith({
        remoteIdentifier: lndPubKey,
        amount: args.amount,
        currency: args.currency,
      });
    });

    test('throws when peer not found', async () => {
      expect.assertions(1);
      service = new Service(components);
      const args = getArgs();
      const peerNotFoundError = new Error('peer not found');
      components.pool.getPeer = jest.fn().mockImplementation(() => {
        throw peerNotFoundError;
      });
      try {
        await service.openChannel(args);
      } catch (e) {
        expect(e).toEqual(peerNotFoundError);
      }
    });

    test('throws when failure from swapClientManager', async () => {
      expect.assertions(1);
      service = new Service(components);
      const args = getArgs();
      components.swapClientManager.openChannel = jest.fn().mockImplementation(() => {
        throw new Error('swapClientManager openChannel failure');
      });
      try {
        await service.openChannel(args);
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe('getBalance', () => {
    const setup = () => {
      service = new Service(components);
      components.swapClientManager.swapClients = new Map();
      components.swapClientManager.get = jest.fn().mockImplementation((arg) => {
        return components.swapClientManager.swapClients.get(arg);
      });

      const btcClient = new mockedSwapClient();
      btcClient.isConnected = jest.fn().mockImplementation(() => true);
      btcClient.channelBalance = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          balance: 70000,
          inactiveBalance: 18817,
          pendingOpenBalance: 190191,
        });
      });
      btcClient.walletBalance = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          totalBalance: 10000,
          confirmedBalance: 10000,
          unconfirmedBalance: 0,
        });
      });
      components.swapClientManager.swapClients.set('BTC', btcClient);

      const ltcClient = new mockedSwapClient();
      ltcClient.isConnected = jest.fn().mockImplementation(() => true);
      ltcClient.channelBalance = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          balance: 0,
          inactiveBalance: 12345,
          pendingOpenBalance: 0,
        });
      });
      ltcClient.walletBalance = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          totalBalance: 2000,
          confirmedBalance: 1500,
          unconfirmedBalance: 500,
        });
      });
      components.swapClientManager.swapClients.set('LTC', ltcClient);

      const bchClient = new mockedSwapClient();
      bchClient.isConnected = jest.fn().mockImplementation(() => false);
      components.swapClientManager.swapClients.set('BCH', bchClient);
    };

    test('returns balance for all connected clients when currency is not specified', async () => {
      setup();
      const result = await service.getBalance({ currency: '' });
      expect(result.size).toEqual(2);

      const btcBalance = result.get('BTC')!;
      expect(btcBalance).toBeTruthy();
      expect(btcBalance.channelBalance).toEqual(70000);
      expect(btcBalance.pendingChannelBalance).toEqual(190191);
      expect(btcBalance.inactiveChannelBalance).toEqual(18817);
      expect(btcBalance.walletBalance).toEqual(10000);
      expect(btcBalance.unconfirmedWalletBalance).toEqual(0);
      expect(btcBalance.totalBalance).toEqual(289008);

      const ltcBalance = result.get('LTC')!;
      expect(ltcBalance).toBeTruthy();
      expect(ltcBalance.channelBalance).toEqual(0);
      expect(ltcBalance.pendingChannelBalance).toEqual(0);
      expect(ltcBalance.inactiveChannelBalance).toEqual(12345);
      expect(ltcBalance.walletBalance).toEqual(1500);
      expect(ltcBalance.unconfirmedWalletBalance).toEqual(500);
      expect(ltcBalance.totalBalance).toEqual(14345);
    });

    test('returns balance for specified currency', async () => {
      setup();
      const result = await service.getBalance({ currency: 'BTC' });
      expect(result.size).toEqual(1);

      const btcBalance = result.get('BTC')!;
      expect(btcBalance).toBeTruthy();
      expect(btcBalance.channelBalance).toEqual(70000);
      expect(btcBalance.pendingChannelBalance).toEqual(190191);
      expect(btcBalance.inactiveChannelBalance).toEqual(18817);
      expect(btcBalance.walletBalance).toEqual(10000);
      expect(btcBalance.unconfirmedWalletBalance).toEqual(0);
      expect(btcBalance.totalBalance).toEqual(289008);
    });

    test('throws in case of invalid currency', async () => {
      setup();
      await expect(service.getBalance({ currency: 'A' })).rejects.toMatchSnapshot();
    });

    test('throws when swap client is not found', async () => {
      setup();
      await expect(service.getBalance({ currency: 'BBB' })).rejects.toMatchSnapshot();
    });
  });

  describe('tradingLimits', () => {
    const setup = () => {
      service = new Service(components);
      components.swapClientManager.swapClients = new Map();
      components.swapClientManager.tradingLimits = jest.fn().mockImplementation((currency) => {
        if (currency === 'BTC') {
          return Promise.resolve({
            maxSell: 2000,
            maxBuy: 1500,
          });
        } else if (currency === 'LTC') {
          return Promise.resolve({
            maxSell: 7000,
            maxBuy: 5500,
          });
        } else {
          return Promise.resolve();
        }
      });

      const btcClient = new mockedSwapClient();
      btcClient.isConnected = jest.fn().mockReturnValue(true);
      components.swapClientManager.swapClients.set('BTC', btcClient);

      const ltcClient = new mockedSwapClient();
      ltcClient.isConnected = jest.fn().mockReturnValue(true);
      components.swapClientManager.swapClients.set('LTC', ltcClient);

      const bchClient = new mockedSwapClient();
      bchClient.isConnected = jest.fn().mockImplementation(() => false);
      components.swapClientManager.swapClients.set('BCH', bchClient);
    };

    test('returns trading limits for all connected clients when currency is not specified', async () => {
      setup();
      const result = await service.tradingLimits({ currency: '' });
      expect(result.size).toEqual(2);

      const btcTradingLimits = result.get('BTC')!;
      expect(btcTradingLimits).toBeTruthy();
      expect(btcTradingLimits.maxSell).toEqual(2000);
      expect(btcTradingLimits.maxBuy).toEqual(1500);

      const ltcTradingLimits = result.get('LTC')!;
      expect(ltcTradingLimits).toBeTruthy();
      expect(ltcTradingLimits.maxSell).toEqual(7000);
      expect(ltcTradingLimits.maxBuy).toEqual(5500);
    });

    test('returns trading limits for specified currency', async () => {
      setup();
      const result = await service.tradingLimits({ currency: 'BTC' });
      expect(result.size).toEqual(1);

      const btcTradingLimits = result.get('BTC')!;
      expect(btcTradingLimits).toBeTruthy();
      expect(btcTradingLimits.maxSell).toEqual(2000);
      expect(btcTradingLimits.maxBuy).toEqual(1500);
    });

    test('throws in case of invalid currency', async () => {
      setup();
      await expect(service.tradingLimits({ currency: 'A' })).rejects.toMatchSnapshot();
    });
  });

  describe('listOrders', () => {
    const pairIds = ['BTC/LTC', 'WETH/BTC'];

    const setup = (peersOrders: any, ownOrders: any) => {
      Object.defineProperty(components.orderBook, 'pairIds', {
        value: pairIds,
      });
      components.orderBook.getPeersOrders = jest.fn().mockImplementation(() => peersOrders);
      components.orderBook.getOwnOrders = jest.fn().mockImplementation(() => ownOrders);
      service = new Service(components);
    };

    const createOrder = (price: number, createdAt: number) => {
      return {
        price,
        createdAt,
      };
    };

    test('returns no orders if nothing is listed in the orderbook', () => {
      const peersOrders = { buyArray: [], sellArray: [] };
      const ownOrders = { buyArray: [], sellArray: [] };
      setup(peersOrders, ownOrders);

      const result = service.listOrders({
        pairId: '',
        owner: Owner.Both,
        limit: 0,
        includeAliases: false,
      });
      expect(result.size).toEqual(pairIds.length);
      expect(result.get(pairIds[0])!.buyArray.length).toEqual(0);
      expect(result.get(pairIds[0])!.sellArray.length).toEqual(0);
      expect(result.get(pairIds[1])!.buyArray.length).toEqual(0);
      expect(result.get(pairIds[1])!.sellArray.length).toEqual(0);
    });

    test('returns both own and peer orders for all trading pairs', () => {
      const peersOrders = {
        buyArray: [createOrder(1, 123)],
        sellArray: [createOrder(3, 222), createOrder(1, 999)],
      };
      const ownOrders = {
        buyArray: [createOrder(1, 999)],
        sellArray: [createOrder(2, 123)],
      };
      setup(peersOrders, ownOrders);

      const result = service.listOrders({
        pairId: '',
        owner: Owner.Both,
        limit: 0,
        includeAliases: false,
      });
      expect(result.size).toEqual(pairIds.length);
      expect(result.get(pairIds[0])!.buyArray.length).toEqual(2);
      expect(result.get(pairIds[0])!.sellArray.length).toEqual(3);
      expect(result.get(pairIds[1])!.buyArray.length).toEqual(2);
      expect(result.get(pairIds[1])!.sellArray.length).toEqual(3);
    });

    test('returns both own and peer orders for the specified trading pair', () => {
      const peersOrders = {
        buyArray: [createOrder(1, 123)],
        sellArray: [createOrder(3, 222), createOrder(1, 999)],
      };
      const ownOrders = {
        buyArray: [createOrder(1, 999)],
        sellArray: [createOrder(2, 123)],
      };
      setup(peersOrders, ownOrders);

      const result = service.listOrders({
        pairId: pairIds[0],
        owner: Owner.Both,
        limit: 0,
        includeAliases: false,
      });
      expect(result.size).toEqual(1);
      expect(result.get(pairIds[0])!.buyArray.length).toEqual(2);
      expect(result.get(pairIds[0])!.sellArray.length).toEqual(3);
    });

    test('returns only peer orders', () => {
      const peersOrders = {
        buyArray: [createOrder(1, 123)],
        sellArray: [createOrder(3, 222), createOrder(1, 999)],
      };
      const ownOrders = {
        buyArray: [createOrder(1, 999), createOrder(1, 123)],
        sellArray: [createOrder(2, 123)],
      };
      setup(peersOrders, ownOrders);

      const result = service.listOrders({
        pairId: pairIds[0],
        owner: Owner.Peer,
        limit: 0,
        includeAliases: false,
      });
      expect(result.size).toEqual(1);
      expect(result.get(pairIds[0])!.buyArray.length).toEqual(1);
      expect(result.get(pairIds[0])!.sellArray.length).toEqual(2);
    });

    test('returns only own orders', () => {
      const peersOrders = {
        buyArray: [createOrder(1, 123)],
        sellArray: [createOrder(3, 222), createOrder(1, 999)],
      };
      const ownOrders = {
        buyArray: [createOrder(1, 999), createOrder(1, 123)],
        sellArray: [createOrder(2, 123)],
      };
      setup(peersOrders, ownOrders);

      const result = service.listOrders({
        pairId: pairIds[0],
        owner: Owner.Own,
        limit: 0,
        includeAliases: false,
      });
      expect(result.size).toEqual(1);
      expect(result.get(pairIds[0])!.buyArray.length).toEqual(2);
      expect(result.get(pairIds[0])!.sellArray.length).toEqual(1);
    });

    test('returns limited amount of orders', () => {
      const peersOrders = {
        buyArray: [createOrder(1, 111)],
        sellArray: [createOrder(4, 222), createOrder(5, 999)],
      };
      const ownOrders = {
        buyArray: [createOrder(3, 999), createOrder(2, 123)],
        sellArray: [createOrder(6, 123)],
      };
      setup(peersOrders, ownOrders);

      const result = service.listOrders({
        pairId: pairIds[0],
        owner: Owner.Both,
        limit: 2,
        includeAliases: false,
      });
      expect(result.size).toEqual(1);
      expect(result.get(pairIds[0])!.buyArray.length).toEqual(2);
      expect(result.get(pairIds[0])!.buyArray.some((val) => val.price === 2)).toBeTruthy();
      expect(result.get(pairIds[0])!.sellArray.length).toEqual(2);
      expect(result.get(pairIds[0])!.sellArray.some((val) => val.price === 5)).toBeTruthy();
    });
  });
});
