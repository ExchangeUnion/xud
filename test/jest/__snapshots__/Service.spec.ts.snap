// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service getBalance throws in case of invalid currency 1`] = `
Object {
  "code": "6.1",
  "message": "currency must consist of 2 to 5 upper case English letters or numbers",
}
`;

exports[`Service getBalance throws when swap client is not found 1`] = `
Object {
  "code": "7.1",
  "message": "swapClient for currency BBB not found",
}
`;

exports[`Service openChannel throws when failure from swapClientManager 1`] = `
Object {
  "code": "6.5",
  "message": "failed to open BTC channel for 16000000 with 02f8895eb03c37b2665415be4d83b20228acc0abc55ebf6728565141c66cfc164a: swapClientManager openChannel failure",
}
`;

exports[`Service tradingLimits throws in case of invalid currency 1`] = `
Object {
  "code": "6.1",
  "message": "currency must consist of 2 to 5 upper case English letters or numbers",
}
`;
